flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala

```diff
@@ -1169,6 +1169,94 @@ object AggregateUtil {
+      isDistinct: Boolean,
+      aggregateInputType: Seq[RelDataType],
+      needRetraction: Boolean,
+      tableConfig: TableConfig,
+      isStateBackedDataViews: Boolean = false,
+      index: Int = 0)
+    : (TableAggregateFunction[_, _],
+      TypeInformation[_],
+      Seq[DataViewSpec[_]]) = {
+    // store the aggregate fields of each aggregate function, by the same order of aggregates.
+    // create aggregate function instances by function type and aggregate field data type.
+
+    var accumulatorType: TypeInformation[_] = null
+    val inputType = () => if (aggregateInputType.isEmpty) {
+      throw new TableException("Aggregate fields should not be empty.")
+    } else {
+      aggregateInputType.get(0)
+    }
+
+    val aggregate: TableAggregateFunction[_, _] = aggregateCall match {
+
+      case _: SqlCountAggFunction =>
+        new CountAggFunction
+
+      case collect: SqlAggFunction if collect.getKind == SqlKind.COLLECT =>
+        val agg = new CollectAggFunction(FlinkTypeFactory.toTypeInfo(inputType()))
+        accumulatorType = agg.getAccumulatorType
+        agg
+
+      case udagg: AggSqlFunction =>
+        accumulatorType = udagg.accType
+        udagg.getFunction
+
+      case aggFunction: SqlAggFunction =>
+        createFlinkAggFunction(aggFunction,
+          needRetraction,
+          inputType(),
+          tableConfig)
+    }
+
+    val accSpecs =
+      if (accumulatorType != null) {
+        val (accType, specs) = removeStateViewFieldsFromAccTypeInfo(index,
+          aggregate,
+          accumulatorType,
+          isStateBackedDataViews)
+        if (specs.isDefined) {
+          accumulatorType = accType
+          specs.get
+        } else {
+          Seq()
+        }
+      } else {
+        accumulatorType = getAccumulatorTypeOfAggregateFunction(aggregate)
+        Seq()
+      }
+
+    // create distinct accumulator filter argument
+    if (isDistinct) {
+      // Using Pojo fields for the real underlying accumulator
+      val pojoFields = new util.ArrayList[PojoField]()
+      pojoFields.add(new PojoField(
+        classOf[DistinctAccumulator[_]].getDeclaredField("realAcc"),
+        accumulatorType)
+      )
+      // If StateBackend is not enabled, the distinct mapping also needs
+      // to be added to the Pojo fields.
+      if (!isStateBackedDataViews) {
+
+        val argTypes: Array[TypeInformation[_]] = aggregateInputType
+          .map(FlinkTypeFactory.toTypeInfo).toArray
+
+        val mapViewTypeInfo = new MapViewTypeInfo(
+          new RowTypeInfo(argTypes: _*),
+          BasicTypeInfo.LONG_TYPE_INFO)
+        pojoFields.add(new PojoField(
+          classOf[DistinctAccumulator[_]].getDeclaredField("distinctValueMap"),
+          mapViewTypeInfo)
+        )
+      }
+      accumulatorType = new PojoTypeInfo(classOf[DistinctAccumulator[_]], pojoFields)
+    }
+
+    (aggregate, accumulatorType, accSpecs)
+  }
+
   private def transformToAggregateFunctions(
```

Comment:
We definitely need more documentation for this function. Can you create case classes for the return type of this function for proper naming because `Array[Int]` or `Array[Boolean]` is not very helpful.